packages/tasit-identity-contract/src/GnosisSafeUtils.js

```diff
@@ -0,0 +1,149 @@
+      const sig = signingKey.signDigest(hash);
+      signatures += ethers.utils.joinSignature(sig).slice(2);
+    });
+
+    return signatures;
+  };
+
+  // Based on: https://github.com/gnosis/safe-contracts/blob/102e632d051650b7c4b0a822123f449beaf95aed/test/utilsPersonalSafe.js#L30
+  estimateFromSafeTxGas = async (to, value, data, operation) => {
+    const provider = this.#contract._getProvider();
+    const safeContractAddress = this.#contract.getAddress();
+
+    // Encode call to the contract's estimate gas function
+    const callToEstimate = this.encodeFunctionCall(
+      this.#contract,
+      "requiredTxGas",
+      [to, value, data, operation]
+    );
+
+    // The response will be return via error message
+    const estimateResponse = await provider.call({
+      to: safeContractAddress,
+      from: safeContractAddress,
+      data: callToEstimate,
+    });
+
+    // Extract estimation from revert error message
+    const estimatedGasHex = "0x" + estimateResponse.substring(138);
+    let estimatedGas = bigNumberify(estimatedGasHex);
+
+    // Extracted from Gnosis Safe test case (Why it's necessary?)
+    // Add 10k else we will fail in case of nested calls
+    estimatedGas = estimatedGas.add(10000);
+
+    return estimatedGas;
+  };
+
+  // Based on: https://github.com/gnosis/safe-contracts/blob/102e632d051650b7c4b0a822123f449beaf95aed/test/utilsPersonalSafe.js#L7
+  estimateDataGas = (
+    safe,
+    to,
+    value,
+    data,
+    operation,
+    txGasEstimate,
+    gasToken,
+    refundReceiver,
+    signatureCount
+  ) => {
+    const sigRcost = 2176;
+    const sigScost = 2176;
+    const sigVcost = 68;
+    const eachSigCost = sigRcost + sigScost + sigVcost;
+    const signatureCost = signatureCount * eachSigCost;
+
+    const dataGas = 0;
+    const gasPrice = GAS_PRICE;
+
+    // Signatures cost will add later
+    const signatures = "0x";
+
+    const callToEstimate = this.encodeFunctionCall(
+      this.#contract,
+      "execTransaction",
+      [
+        to,
+        value,
+        data,
+        operation,
+        txGasEstimate,
+        dataGas,
+        gasPrice,
+        gasToken,
+        refundReceiver,
+        signatures,
+      ]
+    );
+
+    let estimatedGas =
+      this.#estimateDataGasCosts(callToEstimate) + signatureCost;
+
+    if (estimatedGas > 65536) {
+      estimatedGas += 64;
+    } else {
+      estimatedGas += 128;
+    }
+
+    // Extracted from Gnosis Safe test case (Why it's necessary?)
+    // Add aditional gas costs (e.g. base tx costs, transfer costs)
+    estimatedGas += 32000;
+
+    return estimatedGas;
+  };
+
+  // Calculate data gas
+  #estimateDataGasCosts = dataString => {
+    const reducer = (accumulator, currentValue) =>
+      (accumulator += this.#dataGasValue(currentValue));
+
+    return dataString.match(/.{2}/g).reduce(reducer, 0);
+  };
+
+  // Calculate gas for each byte
+  // Note: No references found
```

Comment:
@marcelomorgado Feel free to ask the Gnosis team in our shared Slack channel

Reply:
Perhaps it's related to https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/cryptography/ECDSA.sol ?

Reply:
Extracted to: https://github.com/tasitlabs/TasitSDK/issues/237

Reply:
Thanks. Resolving